Add PushSecret UpdatePolicy (to replace PR #3100)

maresac opened the pull request:
## Problem Statement

Implementation of [PushSecret Update Policy](https://github.com/external-secrets/external-secrets/blob/main/design/008-pushsecret-update-policy.md).

This is a draft version, currently only implemented for the Vault provider.

I'll add documentation after feedback.

-- Disclaimer: I opened this PR as a replacement for #3100 because of merge conflicts I couldn't resolve easily --  

## Related Issue

https://github.com/external-secrets/external-secrets/issues/2643

## Proposed Changes

I mostly stuck to the [design doc](https://github.com/maresac/external-secrets/blob/main/design/008-pushsecret-update-policy.md) with the exception of the PushSecret Status.

My proposal for the Status is the following:
* Set Status to `SecretSynced` if one or more existing values in provider were NOT overwritten, as this is intended behavior.
* Indicate that there are unchanged existing secrets in the provider through event message.

## Checklist

- [ ] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [ ] All commits are signed with `git commit --signoff`
- [ ] My changes have reasonable test coverage
- [ ] All tests pass with `make test`
- [ ] I ensured my PR is ready for review with `make reviewable`

Skarlso commented on apis/externalsecrets/v1alpha1/pushsecret_types.go:
```diff
@@ -54,6 +62,10 @@ type PushSecretSpec struct {
 	// The Interval to which External Secrets will try to push a secret definition
 	RefreshInterval *metav1.Duration     `json:"refreshInterval,omitempty"`
 	SecretStoreRefs []PushSecretStoreRef `json:"secretStoreRefs"`
+	// Update Policy to handle Secrets in the provider. Possible Values: "Replace/IfNotExists". Defaults to "Replace".
```
```suggestion
	// UpdatePolicy to handle Secrets in the provider. Possible Values: "Replace/IfNotExists". Defaults to "Replace".
```

Skarlso commented on pkg/controllers/pushsecret/pushsecret_controller.go:
```diff
@@ -268,34 +289,50 @@ func (r *Reconciler) DeleteSecretFromStore(ctx context.Context, client v1beta1.S
 
+			switch ps.Spec.UpdatePolicy {
+			case esapi.PushSecretUpdatePolicyIfNotExists:
+				exists, err := secretClient.SecretExists(ctx, data.Match.RemoteRef)
```
```suggestion
				exists, err := secretClient.SecretExists(ctx, data.GetSecretKey())
```

Skarlso commented:
Generally, I think this is a good approach. I can't help but wonder though if we could pull these two maps together. 🤔 

But meh, it's fine, I guess. :)

Skarlso commented:
@maresac Can I ask you if you could take care of the sonar issues in pushsecret please? Since you added the new things, they show up as new violations.

Just the push secret one. And there is like one in the push secret controller file ( complexity ) and one reused value in the test. The rest, leave those. :) 

Thanks, after that I think this is good to merge. :)

maresac commented:
Hey @Skarlso, good point. I'll take care of those!
I also thought a little more about pulling the maps together. Initially I didn't do that because I wanted to be as minimally invasive as possible, but I get your point. It could make the status look more concise. I'll work on something.

Skarlso commented:
Hey @maresac.

Do you need any help? :)

maresac commented:
Hey @Skarlso! I just rebased and added missing methods to the new providers, now the pipeline jobs should go through. Sorry, I didn't look closely enough that that was an issue :D

Skarlso commented:
/ok-to-test sha=1da7f79

Skarlso commented:
I'm sorry for asking this, but since you touched the files would you mind just taking care of the 8 sonar issues? So just create a constans for the string "not implemented" and that's it. :D 🙇 🙇 

Then we can get this baby on the road. :D

Skarlso reviewed (approved):
After "not implemented" is in a constant, LGTM

maresac commented:
Don't be sorry @Skarlso, you're right about this :D

Skarlso commented:
Thank you for your work on this @maresac! :)

maresac commented:
Thanks for bearing with me ;)
